### PR TITLE
OCPBUGS-27852: [release-4.15] Full implementation of KEP-1669 ProxyTerminatingEndpoints + ETP=local fix

### PR DESCRIPTION
**Pull request by ricky-rav**
- clean cherry pick of first commit (32d2ce19dd0f9e76eb2a1ca5b966d02507779f0e) from master branch
- cherry pick of ETP=local fix (e2faa152408d9cad1f6cc8dcf0318ab9f4ae6d8c) from master branch, applying the exact same changes we have on master for the affected OCP hacks. I had to replace all calls to `k8s.io/utils/ptr` (e.g. `ptr.To(true)`) with calls to the older library `k8s.io/utils/pointer` (e.g. `utilpointer.Bool(true)`), since that's new code that we only vendor in 4.16.

**First commit:**
The previous implementation was an approximation of KEP-1669 ProxyTerminatingEndpoints: we simply included terminating serving endpoints (ready=false, serving=true, terminating=true) along with ready ones in the endpoint selection logic. Let's fully implement KEP-1669 and only include terminating endpoints if none are ready. The selection follows two simple steps: 1) Take all ready endpoints
2) If no ready endpoints were found, take all serving terminating endpoints.

This should also help with an issue found in a production cluster (https://issues.redhat.com/browse/OCPBUGS-24363) where, due to infrequent readiness probes, terminating endpoints were declared as non-serving (that is, their readiness probe failed) only quite late and were included as valid endpoints for quite a bit, while the existing ready endpoints should have been preferred.

Extended the test cases to include testing against multiple slices and dual stack scenarios.

Signed-off-by: Riccardo Ravaioli <rravaiol@redhat.com>
(cherry picked from commit 418043c1f528213ff2f1afabc4cec66c73f7b815)

**Second commit:**
Fix endpoint selection for externalTrafficPolicy=local
Fix the case for "all endpoints terminating on a node when traffic policy is local":
"When the traffic policy is "Local" and all endpoints are terminating within a single node, then traffic should be routed to any terminating endpoint that is ready on that node."
https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/1669-proxy-terminating-endpoints/README.md#example-all-endpoints-terminating-on-a-node-when-traffic-policy-is-local

The endpoint selection logic in the services controller is now entirely implemented in getEndpointsForService, which computes for a given service and each service port all its cluster-wide endpoints and per-node local endpoints. We first apply a cluster-wide vs local endpoint selection and only then we apply readiness-based filtering with getEligibleEndpointAddresses.

Added unit tests for the new logic.

### CODE REVIEW

**Comment by ricky-rav** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-e2e-gcp-ovn-upgrade 10

**Comment by ricky-rav** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-gcp-ovn-rt-upgrade 10

**Review by tssurya — APPROVED** [review]
/lgtm
exactly same as https://github.com/openshift/ovn-kubernetes/commit/418043c1f528213ff2f1afabc4cec66c73f7b815
@ricky-rav : why do you have hold on this PR? I won't add the risk-assessed until we know we want to merge this. Did you see some diruptions ? Because I saw you triggered the payload job..

**Comment by tssurya** [comment]
self-note: spoke to ricky, he will add more stuff into this fix before bringing it down

**Comment by ricky-rav** [comment]
The case of externalTrafficPolicy=local was not correctly handled in ovnkube-controller code, since it applied readiness selection first and only afterwards it filtered out non-local endpoints.  The reverse order is the correct one and I needed ovnkube-controller to be exposed to endpoints  instead of just endpoint addresses in order to apply the readiness selection in the end and not at the beginning:
https://github.com/openshift/ovn-kubernetes/pull/2075

**Comment by ricky-rav** [comment]
`e2e-metal-ipi-ovn-dualstack` is failing exactly as a no-op PR on 4.15:
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_ovn-kubernetes/2133/pull-ci-openshift-ovn-kubernetes-release-4.15-e2e-metal-ipi-ovn-dualstack/1781004220237877248

File: go-controller/pkg/util/kube.go [hunk 1]
```diff
@@ -878,3 +794,15 @@ func IsHostEndpoint(endpointIPstr string) bool {
...
+	endpoints := []discovery.Endpoint{}
+	for _, slice := range endpointSlices {
+		endpoints = append(endpoints, slice.Endpoints...)
+	}
+	return endpoints
+}
+
+func GetConntrackZone() int {
+	return config.Default.ConntrackZone
+}
```
Comment (jcaamano) [inline go-controller/pkg/util/kube.go:808]: Did you pick this up intentionally?

**Comment by jcaamano** [comment]
/payload periodic-ci-openshift-release-master-ci-4.15-e2e-gcp-ovn-upgrade

**Comment by jcaamano** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-e2e-gcp-ovn-upgrade

**Comment by jcaamano** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-e2e-gcp-ovn-upgrade

**Comment by jcaamano** [comment]
/payload-abort
these are good

**Pull request merged**